osquery/filesystem/windows/fileops.cpp

```diff
@@ -1082,6 +1082,79 @@ size_t PlatformFile::size() const {
   return ::GetFileSize(handle_, nullptr);
 }
 
+bool platformSetSafeDbPerms(const std::string& path) {
+  unsigned long sid_size = SECURITY_MAX_SID_SIZE;
+  std::vector<char> admins_buf(sid_size);
+  PSID admins = static_cast<PSID>(admins_buf.data());
+  if (!::CreateWellKnownSid(
+          WinBuiltinAdministratorsSid, nullptr, admins, &sid_size)) {
+    return false;
+  }
+
+  std::vector<char> system_buf(sid_size);
+  PSID system = static_cast<PSID>(system_buf.data());
+  if (!::CreateWellKnownSid(WinLocalSystemSid, nullptr, system, &sid_size)) {
+    return false;
+  }
+
+  EXPLICIT_ACCESS ea;
+  ea.grfAccessMode = SET_ACCESS;
+  ea.grfAccessPermissions = GENERIC_ALL;
+  ea.grfInheritance = SUB_CONTAINERS_AND_OBJECTS_INHERIT;
+
+  PTRUSTEE_A trust = static_cast<PTRUSTEE_A>(malloc(sizeof(TRUSTEE_A)));
+  BuildTrusteeWithSidA(trust, admins);
+  ea.Trustee = *trust;
+
+  // Set the Administrators ACE
+  PACL new_dacl = nullptr;
+  auto ret = SetEntriesInAcl(1, &ea, nullptr, &new_dacl);
+  if (ret != ERROR_SUCCESS) {
+    VLOG(1) << "Failed to set DB permissions for Administrators";
+    LocalFree(new_dacl);
+    free(trust);
+    return false;
+  }
+
+  // Set the SYSTEM ACE
+  BuildTrusteeWithSidA(trust, system);
+  ea.Trustee = *trust;
+  ret = SetEntriesInAcl(1, &ea, new_dacl, &new_dacl);
+  if (ret != ERROR_SUCCESS) {
+    VLOG(1) << "Failed to set DB permissions for SYSTEM";
+    LocalFree(new_dacl);
+    free(trust);
+    return false;
+  }
+
+  // First clear out the DACL
+  ret = SetNamedSecurityInfoA(const_cast<char*>(path.c_str()),
+                              SE_FILE_OBJECT,
+                              DACL_SECURITY_INFORMATION,
+                              nullptr,
+                              nullptr,
+                              nullptr,
+                              nullptr);
+  if (ret != ERROR_SUCCESS) {
+    VLOG(1) << "Failed to clear DACL for DB permissions";
+  }
+
+  // Apply 'safe' DACL and avoid returning to attempt applying the DACL
+  ret = SetNamedSecurityInfoA(const_cast<char*>(path.c_str()),
+                              SE_FILE_OBJECT,
+                              DACL_SECURITY_INFORMATION,
+                              nullptr,
+                              nullptr,
+                              new_dacl,
```

Comment:
Do we need to LocalFree(new_dacl); before new_dacl reuse?

Reply:
I don't think so, the `LocalFree(new_dacl)`s above were only because the MSDN explicitly asks one to use it when finished with a DACL object, but honestly a `LocalFree` is similar to just destructing stack values and I'm not entirely sure that it's even needed.